File: android/besitos-android-sdk/src/main/kotlin/io/github/kashkick/besitos/android/sdk/offerwall/url/UrlBuilder.kt
package io.github.kashkick.besitos.android.sdk.offerwall.url

import io.github.kashkick.besitos.android.sdk.offerwall.SdkConfig
import io.github.kashkick.besitos.android.sdk.offerwall.config.OfferwallConfig
import java.net.URLEncoder

internal object UrlBuilder {

    fun build(config: OfferwallConfig): String {
        val params = buildList {
            add("userid" to config.userId)
            config.subId?.takeIf { it.isNotBlank() }?.let { add("subid" to it) }
            config.deviceId?.takeIf { it.isNotBlank() }?.let { add("device_id" to it) }
            config.idfa?.takeIf { it.isNotBlank() }?.let { add("idfa" to it) }
            config.gdfa?.takeIf { it.isNotBlank() }?.let { add("gdfa" to it) }
            config.info?.takeIf { it.isNotBlank() }?.let { add("info" to it) }
            if (config.hideHeader) add("hide_header" to "1")
            if (config.hideFooter) add("hide_footer" to "1")
        }

        val query = params.joinToString("&") { (k, v) ->
            "${encode(k)}=${encode(v)}"
        }

        val urlPath = if (config.onlyOffers) "offers" else "v2"
        return "${SdkConfig.BASE_URL}/$urlPath/${encode(config.partnerId)}?$query"
    }

    private fun encode(value: String): String =
        URLEncoder.encode(value, "UTF-8").replace("+", "%20")
}

File: ios/Sources/Offerwall/UrlBuilder.swift
import Foundation

struct UrlBuilder {

    static func build(config: OfferwallConfig) -> URL? {
        let urlPath = config.onlyOffers ? "offers" : "v2"
        var components = URLComponents(string: "\(SdkConfig.baseUrl)/\(urlPath)/\(encode(config.partnerId))")
        var items: [URLQueryItem] = []

        items.append(URLQueryItem(name: "userid", value: config.userId))

        if let subId = config.subId, !subId.isEmpty {
            items.append(URLQueryItem(name: "subid", value: subId))
        }
        if let deviceId = config.deviceId, !deviceId.isEmpty {
            items.append(URLQueryItem(name: "device_id", value: deviceId))
        }
        if let idfa = config.idfa, !idfa.isEmpty {
            items.append(URLQueryItem(name: "idfa", value: idfa))
        }
        if let gdfa = config.gdfa, !gdfa.isEmpty {
            items.append(URLQueryItem(name: "gdfa", value: gdfa))
        }
        if let info = config.info, !info.isEmpty {
            items.append(URLQueryItem(name: "info", value: info))
        }
        items.append(URLQueryItem(name: "hide_header", value: config.hideHeader ? "1" : "0"))
        items.append(URLQueryItem(name: "hide_footer", value: config.hideFooter ? "1" : "0") )

        components?.queryItems = items
        return components?.url
    }

    private static func encode(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? value
    }
}
